inferno/dataset.py

```diff
@@ -0,0 +1,136 @@
+from functools import partial
+
+from sklearn.utils import safe_indexing
+import torch
+
+from inferno.utils import to_tensor
+
+
+def _apply_to_data(data, func, unpack_dict=False):
+    _apply = partial(_apply_to_data, func=func, unpack_dict=unpack_dict)
+    if isinstance(data, dict):
+        if unpack_dict:
+            return [_apply(v) for v in data.values()]
+        else:
+            return {k: _apply(v) for k, v in data.items()}
+    elif isinstance(data, (list, tuple)):
+        try:
+            # e.g.list/tuple of arrays
+            return [_apply(x) for x in data]
+        except TypeError:
+            return func(data)
+    return func(data)
+
+
+def _flatten(arr):
+    for item in arr:
+        if isinstance(item, (tuple, list)):
+            yield from _flatten(item)
+        else:
+            yield item
+
+
+def get_len(data):
+    lens = [_apply_to_data(data, len, unpack_dict=True)]
+    lens = list(_flatten(lens))
+    len_set = set(lens)
+    if len(len_set) != 1:
+        raise ValueError("Dataset does not have consistent lengths.")
+    return list(len_set)[0]
+
+
+def multi_indexing(data, i):
+    if isinstance(data, dict):
+        # dictionary of containers
+        return {k: v[i] for k, v in data.items()}
+    if isinstance(data, (list, tuple)):
+        # list or tuple of containers
+        try:
+            return [multi_indexing(x, i) for x in data]
+        except TypeError:
+            pass
+    if torch.is_tensor(data):
+        # torch tensor-like
+        return data[i]
+    if hasattr(data, 'iloc'):
+        # pandas NDFrame
+        return data.iloc[i]
+    # numpy ndarray, list
+    if isinstance(i, (int, slice)):
+        return data[i]
+    return safe_indexing(data, i)
+
+
+class Dataset(object):
+    """General dataset wrapper that can be used in conjunction with
+    pytorch's DataLoader.
+
+    The dataset will always yield two values, the data and the
```

Comment:
"The dataset will always yield two values, the data (`X`) and the target (`y`). [...]"